### PR TITLE
Refactor Base58 tests

### PR DESCRIPTION
**Pull request by MethDamon**

### CODE REVIEW

**Comment by schildbach** [comment]
Can you elaborate a bit on how you refactored? Do the tests cover everything they did before?

**Comment by MethDamon** [comment]
**testEncode() refactored to Base58EncodeTest**
- 	encode("Hello World".getBytes()) -> "JxF12TrwUP45BMd"
- 	encode(BigInteger.valueOf(3471844090L)).toByteArray() -> "16Ho7Hs"
- 	encode(new byte[1]) -> "1"
- 	encode(new byte[7]) -> "1111111"
- 	encode(new byte[0]) -> ""

**testEncodeChecked_address() and testEncodeChecked_privateKey() refactored to Base58EncodeCheckedTest.java**
- 	encodeChecked(111, new byte[LegacyAddress.LENGTH]) -> "mfWxJ45yp2SFn7UciZyNpvDKrzbhyfKrY8"
- 	encodeChecked(128, new byte[32]) -> "5HpHagT65TZzG1PH3CSu63k8DbpvD8s5ip4nEB3kEsreAbuatmU"

**testDecode() testDecode_invalidBase58() refactored to Base58DecodeTest.java and added testDecode_emptyString()**
- 	decode("JxF12TrwUP45BMd") -> "Hello World".getBytes() -> new String(decode("JxF12TrwUP45BMd")))
- 	decode("1") -> new byte[1] -> "1"
- 	decode("1111") -> new byte[4] -> "1111"

	
**testDecodeChecked(), , decode_invalidCharacter_notInAlphabet, testDecodeChecked_shortInput() refactored to Base58DecodeCheckedTest.java**
- 	decodeChecked("4stwEBjT6FYyVV") -> works
- 	decodeChecked("93VYUMzRG9DdbRP72uQXjaWibbQwygnvaCu9DumcqDjGybD864T") -> works
- 	decodeChecked("J0F12TrwUP45BMd") -> AddressFormatException.InvalidCharacter.class
- 	decodeChecked("4s") -> AddressFormatException.InvalidDataLength.class

**testDecodeChecked_invalidChecksum() refactored to Base58DecodeCheckedInvalidChecksum.java**
- 	decodeChecked("4stwEBjT6FYyVW") -> AddressFormatException.InvalidChecksum.class

	

**testDecodeToBigInteger() refactored to Base58DecodeToBigIntegerTest.java()**

- 	decode("129") -> new BigInteger(1, input)

File: core/src/test/java/org/bitcoinj/core/Base58DecodeCheckedTest.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    };
+
+    @Theory
+    public void testDecodeChecked(String input) {
+        Assume.assumeTrue(containsOnlyValidBase58Chars(input));
+        Assume.assumeTrue(input.length() > 4);
+        Base58.decodeChecked(input);
+    }
+
+    @Theory()
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/core/Base58DecodeCheckedTest.java]: Why the "()" here?

File: core/src/test/java/org/bitcoinj/core/Base58DecodeTest.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    @Test
+    public void testDecode_emptyString() {
+        assertEquals(0, Base58.decode("").length);
+    }
+
+    @Test(expected = AddressFormatException.class)
+    public void testDecode_invalidBase58() {
+        Base58.decode("This isn't valid base58");
+    }
+
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/core/Base58DecodeTest.java]: Nitpick: superfluous new line.

File: core/src/test/java/org/bitcoinj/core/Base58DecodeTest.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        });
+    }
+
+    @Test
+    public void testDecode() {
+        byte[] actualBytes = Base58.decode(input);
+        assertArrayEquals(input,  actualBytes, expected);
+    }
+
+    @Test
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/core/Base58DecodeTest.java]: This test (and the next) run 3 times though it doesn't use the different input data.

File: core/src/test/java/org/bitcoinj/core/Base58DecodeCheckedInvalidChecksum.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package org.bitcoinj.core;
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/core/Base58DecodeCheckedInvalidChecksum.java]: All the new files are missing the license. You can use the "Copyright by the original author or authors." copyright line if you want.

File: core/src/test/java/org/bitcoinj/core/Base58DecodeTest.java (hunk 3 above)
Comment (MethDamon) [reply core/src/test/java/org/bitcoinj/core/Base58DecodeTest.java]: True. Do you want me to add two seperate test classes for those or refactor the test using a theory, then I can include an assumption about the string being valid base58 or an empty string.

File: core/src/test/java/org/bitcoinj/core/Base58DecodeCheckedTest.java (hunk 1 above)
Comment (MethDamon) [reply core/src/test/java/org/bitcoinj/core/Base58DecodeCheckedTest.java]: Fixed

File: core/src/test/java/org/bitcoinj/core/Base58DecodeTest.java (hunk 2 above)
Comment (MethDamon) [reply core/src/test/java/org/bitcoinj/core/Base58DecodeTest.java]: Fixed, also in the other files affected by the PR.

File: core/src/test/java/org/bitcoinj/core/Base58DecodeCheckedInvalidChecksum.java (hunk 4 above)
Comment (MethDamon) [reply core/src/test/java/org/bitcoinj/core/Base58DecodeCheckedInvalidChecksum.java]: I added the same license as in the other test files.

**Comment by MethDamon** [comment]
@schildbach Can you take another look at this?

**Comment by MethDamon** [comment]
@schildbach Bump :D I looked at the output of the Travis build but unfortunately I don't really know enough about the project to see why the test is failing, could you take a look at this? Thanks!

File: core/src/test/java/org/bitcoinj/core/Base58DecodeTest.java (hunk 3 above)
Comment (schildbach) [reply core/src/test/java/org/bitcoinj/core/Base58DecodeTest.java]: If there is an easy way to keep the test in the same class then I think I'd prefer it that way.

**Comment by schildbach** [comment]
I restarted the failing test, it's all fine now.

File: core/src/test/java/org/bitcoinj/core/Base58DecodeTest.java (hunk 3 above)
Comment (MethDamon) [reply core/src/test/java/org/bitcoinj/core/Base58DecodeTest.java]: I don't think there is with Parametrized tests. The problem with using Theories is that I cannot easily include the expected values to compare to the inputs used.

Comment (schildbach) [reply core/src/test/java/org/bitcoinj/core/Base58DecodeTest.java]: Ok, then let's keep it as is.

**Comment by schildbach** [comment]
Merged! Thanks for your effort and for your patience.

File: core/src/test/java/org/bitcoinj/core/Base58DecodeTest.java (hunk 2 above)
Comment (schildbach) [reply core/src/test/java/org/bitcoinj/core/Base58DecodeTest.java]: btw. The newlines at EOF are actually wanted. My above comment was about a superfluous line inbetween.

**Comment by MethDamon** [comment]
Cool, thanks for letting me contribute and learn!
